steering_controllers_library/src/steering_odometry.cpp

```diff
@@ -0,0 +1,341 @@
+/*********************************************************************
+* Copyright (c) 2023, Stogl Robotics Consulting UG (haftungsbeschränkt)
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*   http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+ *********************************************************************/
+
+/*
+ * Author: dr. sc. Tomislav Petkovic
+ * Author: Dr. Ing. Denis Stogl
+ */
+
+#include "steering_controllers_library/steering_odometry.hpp"
+
+#include <cmath>
+#include <iostream>
+
+namespace steering_odometry
+{
+SteeringOdometry::SteeringOdometry(size_t velocity_rolling_window_size)
+: timestamp_(0.0),
+  x_(0.0),
+  y_(0.0),
+  heading_(0.0),
+  linear_(0.0),
+  angular_(0.0),
+  wheel_track_(0.0),
+  wheelbase_(0.0),
+  wheel_radius_(0.0),
+  traction_wheel_old_pos_(0.0),
+  velocity_rolling_window_size_(velocity_rolling_window_size),
+  linear_acc_(velocity_rolling_window_size),
+  angular_acc_(velocity_rolling_window_size)
+{
+}
+
+void SteeringOdometry::init(const rclcpp::Time & time)
+{
+  // Reset accumulators and timestamp:
+  reset_accumulators();
+  timestamp_ = time;
+}
+
+bool SteeringOdometry::update_odometry(
+  const double linear_velocity, const double angular, const double dt)
+{
+  /// Integrate odometry:
+  SteeringOdometry::integrate_exact(linear_velocity * dt, angular);
+
+  /// We cannot estimate the speed with very small time intervals:
+  if (dt < 0.0001)
+  {
+    return false;  // Interval too small to integrate with
+  }
+
+  /// Estimate speeds using a rolling mean to filter them out:
+  linear_acc_.accumulate(linear_velocity);
+  angular_acc_.accumulate(angular / dt);
+
+  linear_ = linear_acc_.getRollingMean();
+  angular_ = angular_acc_.getRollingMean();
+
+  return true;
+}
+
+bool SteeringOdometry::update_from_position(
+  const double traction_wheel_pos, const double steer_pos, const double dt)
```

Comment:
they totally should

Reply:
No, they shouldn't. Because it depends on the wheels and configuration, how optometry is calculated. And we need support for the calculating odometry when we have only position interfaces towards hardware.

Reply:
I addressed this